python/ray/train/_internal/storage.py

```diff
@@ -168,11 +171,17 @@ def _download_from_fs_path(
         _local_path.parent.mkdir(parents=True, exist_ok=True)
 
     try:
+        # Note that use_threads=True is safe for download, just not for uploads, see:
+        # https://github.com/apache/arrow/issues/32372
         if filelock:
             with TempFileLock(f"{os.path.normpath(local_path)}.lock"):
-                _pyarrow_fs_copy_files(fs_path, local_path, source_filesystem=fs)
+                _pyarrow_fs_copy_files(
+                    fs_path, local_path, source_filesystem=fs, use_threads=True
+                )
         else:
-            _pyarrow_fs_copy_files(fs_path, local_path, source_filesystem=fs)
+            _pyarrow_fs_copy_files(
+                fs_path, local_path, source_filesystem=fs, use_threads=True
+            )
```

Comment:
Can we change this here instead: https://github.com/ray-project/ray/blob/8a36db78fcfd9ba5a20285f013f05b235dadbc4c/python/ray/train/_internal/storage.py#L96-L101

Condition should just be if `destination_filesystem == S3`

Reply:
That actually would be incorrect--- we have to set it when performing a read only, not for writes (which hang).

The default is True anyways, so it's OK to always set it unconditionally for reads without checking the underlying filesystem type.

Reply:
Oh, I just mean to set it in `_pyarrow_copy_files` rather than `_download_from_fs_path`:

```python
 if isinstance(destination_filesystem, pyarrow.fs.S3FileSystem):
     # Workaround multi-threading issue with pyarrow 
     # https://github.com/apache/arrow/issues/32372 
     kwargs.setdefault("use_threads", False) 
```

Otherwise doesn't the kwarg get overriden as `False`?

Reply:
Right, we cannot set it in pyarrow_copy_files because that's used for both upload and download. It has to be set during the download function only.

Reply:
If it's an upload to s3, won't the destination filesystem check be enough? I'm just wondering if there's any need to check if source_filesystem == s3 anymore. That check just confuses me right now, since it seems to suggest: no threads ever if an s3 fs is present.

Reply:
Oh I see, I missed that you were changing the if conditional and not the body, I'll make that change.